### PR TITLE
[Feature]: export onnx model to support dynamic input size

### PR DESCRIPTION
**Pull request by tangyanf**
This PR convert the calculations related to input size into onnx op, no longer constant tensor, so that the exported onnx model can support dynamic input size.

### CODE REVIEW

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 1]
```diff
@@ -199,9 +199,25 @@ def delta2bbox(rois,
     x2 = gx + gw * 0.5
     y2 = gy + gh * 0.5
     if clip_border and max_shape is not None:
-        x1 = x1.clamp(min=0, max=max_shape[1])
-        y1 = y1.clamp(min=0, max=max_shape[0])
-        x2 = x2.clamp(min=0, max=max_shape[1])
-        y2 = y2.clamp(min=0, max=max_shape[0])
+        if torch.onnx.is_in_onnx_export():
+            min = torch.tensor(0, dtype=torch.float32)
```
Comment (RunningLeon) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:203]: Try use meaningful variable name for `min`, `max1`, `max2`. For example: `zero`, `width`, 'height'.

File: mmdet/models/dense_heads/anchor_head.py [hunk 2]
```diff
@@ -510,6 +511,7 @@ def get_bboxes(self,
                 level with shape (N, num_anchors * 4, H, W)
             img_metas (list[dict]): Meta information of each image, e.g.,
                 image size, scaling factor, etc.
+            imgs (list[torch.Tensor]): List of multiple images
```
Comment (RunningLeon) [inline mmdet/models/dense_heads/anchor_head.py:514]: If you only want to use input image shape, better to get the shape as tensor at preprocessing step and feed it to img_metas. Then, there's no change to the interface of `get_bboxes` method.
For example :
```
input_shape_t = one_metas[img_id]['input_shape_t']
```

File: mmdet/models/dense_heads/anchor_head.py [hunk 3]
```diff
@@ -633,7 +639,7 @@ def _get_bboxes_single(self,
                 scores = cls_score.softmax(-1)
             bbox_pred = bbox_pred.permute(1, 2, 0).reshape(-1, 4)
             nms_pre = cfg.get('nms_pre', -1)
-            if nms_pre > 0 and scores.shape[0] > nms_pre:
+            if nms_pre > 0:
```
Comment (RunningLeon) [inline mmdet/models/dense_heads/anchor_head.py:642]: Setting K of TopK to dynamic could be aligned with this [PR #4684](https://github.com/open-mmlab/mmdetection/pull/4684/files) for the following reason:

1. Only create one `nms_pre` constant tensor;
2. No change to original logic for training and testing in PyTorch.

File: mmdet/models/dense_heads/anchor_head.py (hunk 2 above)
Comment (tangyanf) [reply mmdet/models/dense_heads/anchor_head.py:514]: I tried this, but it is still a constant in the model, like this:
image
We should turn it into a calculation related to the input shape，like this:
image

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py (hunk 1 above)
Comment (tangyanf) [reply mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:203]: sure,i'll update it

File: mmdet/models/dense_heads/anchor_head.py (hunk 2 above)
Comment (RunningLeon) [reply mmdet/models/dense_heads/anchor_head.py:514]: That's a bit strange. I'll try it as well.

File: mmdet/models/detectors/single_stage.py [hunk 4]
```diff
@@ -112,7 +112,7 @@ def simple_test(self, img, img_metas, rescale=False):
         x = self.extract_feat(img)
         outs = self.bbox_head(x)
         bbox_list = self.bbox_head.get_bboxes(
-            *outs, img_metas, rescale=rescale)
+            *outs, img_metas, imgs=[img], rescale=rescale)
```
Comment (RunningLeon) [inline mmdet/models/detectors/single_stage.py:115]: Could pass `img_shape` here above line 114:
```
        if torch.onnx.is_in_onnx_export():
            img_metas[0]['img_shape_for_onnx'] = torch._shape_as_tensor(img)[2:]
```

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 5]
```diff
@@ -199,9 +199,25 @@ def delta2bbox(rois,
     x2 = gx + gw * 0.5
     y2 = gy + gh * 0.5
     if clip_border and max_shape is not None:
-        x1 = x1.clamp(min=0, max=max_shape[1])
-        y1 = y1.clamp(min=0, max=max_shape[0])
-        x2 = x2.clamp(min=0, max=max_shape[1])
-        y2 = y2.clamp(min=0, max=max_shape[0])
+        if torch.onnx.is_in_onnx_export():
```
Comment (RunningLeon) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:204]: You may add comments here why to have this change.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by RunningLeon** [comment]
@ZwwWayne See if you have any issue with this PR.

**Review by ZwwWayne — APPROVED** [review]

**Comment by mchalecki** [comment]
This commit causes error during export of mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco.py using supplied checkpoint from zoo (mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth)

```
Traceback (most recent call last):
  File "tools/deployment/pytorch2onnx.py", line 244, in <module>
    cfg_options=args.cfg_options)
  File "tools/deployment/pytorch2onnx.py", line 57, in pytorch2onnx
    opset_version=opset_version)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/__init__.py", line 230, in export
    custom_opsets, enable_onnx_checker, use_external_data_format)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 91, in export
    use_external_data_format=use_external_data_format)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 639, in _export
    dynamic_axes=dynamic_axes)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 411, in _model_to_graph
    use_new_jit_passes)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 379, in _create_jit_graph
    graph, torch_out = _trace_and_get_graph_from_model(model, args)
  File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 342, in _trace_and_get_graph_from_model
    torch.jit._get_trace_graph(model, args, strict=False, _force_outplace=False, _return_inputs_states=True)
  File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 1148, in _get_trace_graph
    outs = ONNXTracedModule(f, strict, _force_outplace, return_inputs, _return_inputs_states)(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 130, in forward
    self._force_outplace,
  File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 116, in wrapper
    outs.append(self.inner(*trace_inputs))
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 725, in _call_impl
    result = self._slow_forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 709, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/mmcv/runner/fp16_utils.py", line 84, in new_func
    return old_func(*args, **kwargs)
  File "/content/mmdetection/mmdet/models/detectors/base.py", line 183, in forward
    return self.forward_test(img, img_metas, **kwargs)
  File "/content/mmdetection/mmdet/models/detectors/base.py", line 160, in forward_test
    return self.simple_test(imgs[0], img_metas[0], **kwargs)
  File "/content/mmdetection/mmdet/models/detectors/two_stage.py", line 194, in simple_test
    proposal_list = self.rpn_head.simple_test_rpn(x, img_metas)
  File "/content/mmdetection/mmdet/models/dense_heads/rpn_test_mixin.py", line 36, in simple_test_rpn
    proposal_list = self.get_bboxes(*rpn_outs, img_metas)
  File "/usr/local/lib/python3.7/dist-packages/mmcv/runner/fp16_utils.py", line 164, in new_func
    return old_func(*args, **kwargs)
  File "/content/mmdetection/mmdet/models/dense_heads/anchor_head.py", line 572, in get_bboxes
    img_shape = img_metas[img_id]['img_shape_for_onnx']
KeyError: 'img_shape_for_onnx'
```

**Comment by tangyanf** [comment]
> This commit causes error during export of mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco.py using supplied checkpoint from zoo (mask_rcnn_r50_caffe_fpn_mstrain-poly_3x_coco_bbox_mAP-0.408__segm_mAP-0.37_20200504_163245-42aa3d00.pth)
> 
> ```
> Traceback (most recent call last):
>   File "tools/deployment/pytorch2onnx.py", line 244, in <module>
>     cfg_options=args.cfg_options)
>   File "tools/deployment/pytorch2onnx.py", line 57, in pytorch2onnx
>     opset_version=opset_version)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/__init__.py", line 230, in export
>     custom_opsets, enable_onnx_checker, use_external_data_format)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 91, in export
>     use_external_data_format=use_external_data_format)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 639, in _export
>     dynamic_axes=dynamic_axes)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 411, in _model_to_graph
>     use_new_jit_passes)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 379, in _create_jit_graph
>     graph, torch_out = _trace_and_get_graph_from_model(model, args)
>   File "/usr/local/lib/python3.7/dist-packages/torch/onnx/utils.py", line 342, in _trace_and_get_graph_from_model
>     torch.jit._get_trace_graph(model, args, strict=False, _force_outplace=False, _return_inputs_states=True)
>   File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 1148, in _get_trace_graph
>     outs = ONNXTracedModule(f, strict, _force_outplace, return_inputs, _return_inputs_states)(*args, **kwargs)
>   File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 727, in _call_impl
>     result = self.forward(*input, **kwargs)
>   File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 130, in forward
>     self._force_outplace,
>   File "/usr/local/lib/python3.7/dist-packages/torch/jit/_trace.py", line 116, in wrapper
>     outs.append(self.inner(*trace_inputs))
>   File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 725, in _call_impl
>     result = self._slow_forward(*input, **kwargs)
>   File "/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py", line 709, in _slow_forward
>     result = self.forward(*input, **kwargs)
>   File "/usr/local/lib/python3.7/dist-packages/mmcv/runner/fp16_utils.py", line 84, in new_func
>     return old_func(*args, **kwargs)
>   File "/content/mmdetection/mmdet/models/detectors/base.py", line 183, in forward
>     return self.forward_test(img, img_metas, **kwargs)
>   File "/content/mmdetection/mmdet/models/detectors/base.py", line 160, in forward_test
>     return self.simple_test(imgs[0], img_metas[0], **kwargs)
>   File "/content/mmdetection/mmdet/models/detectors/two_stage.py", line 194, in simple_test
>     proposal_list = self.rpn_head.simple_test_rpn(x, img_metas)
>   File "/content/mmdetection/mmdet/models/dense_heads/rpn_test_mixin.py", line 36, in simple_test_rpn
>     proposal_list = self.get_bboxes(*rpn_outs, img_metas)
>   File "/usr/local/lib/python3.7/dist-packages/mmcv/runner/fp16_utils.py", line 164, in new_func
>     return old_func(*args, **kwargs)
>   File "/content/mmdetection/mmdet/models/dense_heads/anchor_head.py", line 572, in get_bboxes
>     img_shape = img_metas[img_id]['img_shape_for_onnx']
> KeyError: 'img_shape_for_onnx'
> ``

the two_stage func will support this feature soon, wait a moment

**Pull request merged**